File: resource/src/com/scamall/resource/Provider.java
/**
 * 
 */
package com.scamall.resource;

/** An abstract class for Resource Providers.
 * 
 *  All Providers of resources must extend this class. All Providers must be thread-safe
 * @author David Navarro Estruch
 *
 */
public abstract class Provider {

	/**
	 * Gets the Providers's internal ID.
	 * 
	 * @return The ID.
	 */
	public abstract String getId();

	
	/** Request a resource.
	 * 
	 * @param app The resource owner app
	 * @param key The key that the app will use to refer the resource.
	 * @param args Any number of additional args that a Provider can ask.
	 */
	abstract public void request(String app, String key, String[] args);

	/** Gets the value from a given resource by the Provider.
	 * 
	 * @param key The key that refers to the resource.
	 * @param subkey A subkey for an specific data of the resource.
	 * @return The subkey data of the key resource.
	 */
	abstract public String value(String key, String subkey);
	
	/** Releases a resource. After a release the resource must not be used again.
	 * 
	 * @param key The key that the app use to refer the resource.
	 * @param args Any number of additional args that a Provider can ask.
	 */
	abstract void release(String key, String[] args);
	
}

File: resource/main/src/nublic_resource/provider.py
'''
Created on 10/08/2010

@author: David Navarro Estruch
@copyright: 2011 Nublic
'''
from elixir import setup_all, session

from model import App, Key
class Provider(object):
    '''
    classdocs
    '''

    def __init__(self, providerType):
        '''
        Constructor
        '''
        self.providerType = providerType

    def install(self, app, key, *args):
        '''
        Must perform a install over the given app, key.
        Optionally can have a number of ordered args.
        
        @param app: string The AppId
        @param key: string The wanted key to identify in the future this resource
        @param args: list Optional args
        
        Raises RepeatedKeyError
        '''
        raise NotImplementedError()
    
    def uninstall(self, app, key):
        '''
        Must perform a uninstall over the given app, key.
        Optionally can have a number of ordered args.
        
        @param app: string The AppId
        @param key: string The wanted key to uninstall
        @param args: list Optional args
        
        @raise NotExistingKeyError
        '''
        raise NotImplementedError()

    def request(self, app, key, *args):
        '''
        Must perform a request over the given app, key.
        Optionally can have a number of ordered args.
        
        @param app: string The AppId
        @param key: string The wanted key to identify in the future this request
        @param args: list Optional args
        
        Raises RepeatedKeyError
        '''
        raise NotImplementedError()
    
    def release(self, app, key):
        '''
        Must perform a release over the given app, key.
        Optionally can have a number of ordered args.
        
        @param app: string The AppId
        @param key: string The wanted key to release
        @param args: list Optional args
        
        @raise NotExistingKeyError
        '''
        raise NotImplementedError()
        
    def value(self, app, key, subkey):
        '''
        Returns the value of the given pair key/subkey.
        
        @param app: string The AppId
        @param key: string The key used to Request
        @param subkey string The subkey wanted. If None return the None subkey.
        
        @return: string value asked
        @raise NotExistingKeyError: Contains the key or subkey that does not exist.
        '''
        raise NotImplementedError()

    def available_values(self, app, key):
        '''
        Returns a list of the available subkeys.

        @param app: string The AppId
        @param key: string The key used to Request
        
        @return: List of string with the available subkeys
        '''
        raise NotImplementedError()

def get_all_apps():
    return App.query.all()
    
def get_all_keys():
    return Key.query.all()
            
class ReapeatedKeyError(Exception):
    def __init__(self, key):
        self.key = key
        
class NotExistingKeyError(Exception):
    def __init__(self, key):
        self.key = key
